[mdatagen] Expose ability to see if enabled option is overridden in user settings

dmitryax opened the pull request:
The new API allows developers to show warnings if `enabled` option isn't overridden in user settings as suggested in https://github.com/open-telemetry/opentelemetry-collector/pull/6358 for
- Replacing a default metric
- Adding a new default metric or turning an existing optional metric into default

Aneurysm9 commented on cmd/mdatagen/metrics.tmpl:
```diff
@@ -19,7 +19,16 @@ import (
+}
+
+// IsEnabled returns a boolean flag defining if the metric is enabled by looking at user settings and the default value.
+func (ms MetricSettings) IsEnabled() bool {
```
Should this also add an `IsUserProvided() bool` helper?  It's not hard to check `ms.Enabled != nil` but having a common way of doing it is helpful.

dmitryax replied:
Makes sense. I also realized that using pointers is not the best way. Let me try a custom unmarshaler instead

dmitryax replied:
Custom unmarshaler worked pretty well. @Aneurysm9 PTAL

dmitryax committed:
[mdatagen] Expose ability to see if enabled field is set in settings

This allows receiver developers to see if the default value is overridden in user settings.

codeboten reviewed (approved):
Tests are failing, but pr looks good

```
ProcessRuntimeMemstatsSys:metadata.MetricSettings{Enabled:true, enabledProvidedByUser:false}, ProcessRuntimeMemstatsTotalAlloc:metadata.MetricSettings{Enabled:true, enabledProvidedByUser:true}}}
        	            	
        	            	Diff:
        	            	--- Expected
make[2]: *** [../../Makefile.Common:57: test] Error 1
make[1]: *** [Makefile:188: receiver/expvarreceiver] Error 2
make: *** [Makefile:95: gotest] Error 2
        	            	+++ Actual
        	            	@@ -88,3 +88,3 @@
        	            	    Enabled: (bool) false,
        	            	-   enabledProvidedByUser: (bool) false
        	            	+   enabledProvidedByUser: (bool) true
        	            	   },
        	            	@@ -140,3 +140,3 @@
        	            	    Enabled: (bool) true,
        	            	-   enabledProvidedByUser: (bool) false
        	            	+   enabledProvidedByUser: (bool) true
        	            	   }
```
